### PR TITLE
Use custom address types on internal bridge functions

### PR DESCRIPTION
**Pull request by maurelian**
**Description**

This is a draft PR for general feedback on this approach to adding more type safety to our Bridge contracts. 

I felt that seeing 4 consecutive `address` typed arguments on most of the bridge functions felt dangerous, ie.

```solidity
address _localToken,
address _remoteToken,
address _from,
address _to,
```

And indeed we have had at least [one issue](https://github.com/ethereum-optimism/optimism/commit/98d0e06e28698fbdff74a4f0f98660e750f900fd#diff-b5041e011abda1d1d726b7fd58224f6cdded883611096c7ad26bd35149020593L242-R245) in the past where we mixed them up. My proposal would help to separate these very different address types:

```solidity
LocalAccount _localToken,
RemoteAccount _remoteToken,
LocalAccount _from,
RemoteAccount _to,
```

**Additional context**

1. This PR only applies changes `internal` functions on the bridge. Making the change to `public`/`external` functions will require a much larger diff to update the tests, and might also create additional hassle for external projects that we should consider.
2. I've created some [utility functions](https://github.com/ethereum-optimism/optimism/pull/3064/files#diff-8848a330370de68e062cb6ec2075f3d451755e40f6e14a050c4ed1e4c89f6183) that make the syntax for type casting much more readable, ie.

```
// rather than burying the address variable in the middle of this boilerplate:
LocalAccount.wrap(_from);

// keep the variable at the front of the expression:
_from.toLocal();
```

Even though this would modify code that is frozen for audit, I think we can easily argue that it is strictly an improvement and does not make any risky modifications to logic.

### CODE REVIEW

File: packages/contracts-bedrock/contracts/libraries/TypedAddresses.sol [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+library TypedAddresses {
+    function toAddress(LocalAccount self) internal pure returns (address) {
+        return LocalAccount.unwrap(self);
+    }
+
+    function toAddress(RemoteAccount self) internal pure returns (address) {
+        return RemoteAccount.unwrap(self);
+    }
+
+    function toLocal(address self) internal pure returns (LocalAccount) {
```
Comment (smartcontracts) [inline packages/contracts-bedrock/contracts/libraries/TypedAddresses.sol:17]: Minor preference to `asLocalAccount` and `asRemoteAccount`

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 2]
```diff
@@ -253,10 +275,10 @@ abstract contract StandardBridge is Initializable {
         bytes calldata _extraData
     ) public virtual {
         _initiateBridgeERC20(
-            _localToken,
-            _remoteToken,
-            msg.sender,
-            _to,
+            _localToken.toLocal(),
```
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/StandardBridge.sol:278]: I think it's a good idea to do it. I took this shortcut here just to reduce the size of the changes needed to get a working PoC. 

The one downside is that other projects would also need to do this casting before calling the bridge, but that also has the benefit of 'exporting' this type safety.

**Comment by maurelian** [comment]
The more I think about it the more I like this change for the added safety it will give us in any future refactoring. But I'd like to make sure we're in agreement. 

I think the options are: 

1. Merge as is (plus some clean up), with typed arguments only applied to internal functions. 
2. Expand typed arguments to the external/public interfaces (providing type safety to external projects as well). 
3. Reject this change (the argument being that this detracts from readability).

**Comment by maurelian** [comment]
Closing for now, we can revisit again in the future

**Pull request closed without merging**